File: Programmers/Python3/Level 2/기능개발/solution.py
# https://programmers.co.kr/learn/courses/30/lessons/42586

from collections import deque


def solution(progresses, speeds):
    progresses = deque(progresses)  # deque의 popleft가 리스트의 pop(0)보다 빠르므로 사용
    speeds = deque(speeds)
    answer = []
    count = 0
    time = 0
    while progresses:
        if progresses[0] + time * speeds[0] >= 100:    # 작업완료
            count += 1
            progresses.popleft()
            speeds.popleft()
        else:
            time += 1
            if count > 0:
                answer.append(count)
                count = 0
    answer.append(count)    # 마지막 작업 처리
    return answer


solution([93, 30, 55], [1, 30, 5])

File: Programmers/Javascript/Level 2/기능개발/solution.js
// https://school.programmers.co.kr/learn/courses/30/lessons/42586?language=javascript

function solution(progresses, speeds) {
  const answer = [];

  while (progresses.length > 0) {
    // 첫번째 작업이 완료될때까지 걸리는 날짜
    const firstProgressDayCount = Math.ceil((100 - progresses[0]) / speeds[0]);

    for (let idx = 0; idx < progresses.length; idx++) {
      progresses[idx] += firstProgressDayCount * speeds[idx];
    }

    // 첫번째 작업과 같이 배포할수있는 작업개수를 구함
    let doneProgresses = 0;
    for (let idx = 0; idx < progresses.length; idx++) {
      if (progresses[idx] < 100) {
        break;
      }
      doneProgresses += 1;
    }

    progresses = progresses.slice(doneProgresses);
    speeds = speeds.slice(doneProgresses);
    answer.push(doneProgresses);
  }
  return answer;
}
